Revamp how we handle Prometheus metrics

lutter opened the pull request:
The biggest change in this PR is that metrics no longer have dynamically constructed names; instead, names are fixed, and labels are used to attach, e.g. the subgraph to the metric.

The PR also changes the `MetricsRegistry` trait a bit to make the various constructor methods a little more ergonomic and clearer in what they should be used for (like `new_counter` vs `new_subgraph_counter`)

Jannis commented on graph/src/components/metrics/mod.rs:
```diff
@@ -11,57 +11,203 @@ pub mod stopwatch;
+                .collect::<Vec<&str>>()
+                .as_slice(),
+        )?);
+        self.register(name, histograms.clone());
```
Have you tested this with multiple subgraphs? I'm wondering if registering the same name twice won't fail.

lutter replied:
I have now run it in the integration cluster, and that worked without errors. I'll let it run over the weekend, but it looks like (after the additions I made) it's all working.

lutter commented on chain/ethereum/src/network_indexer/block_writer.rs:
```diff
@@ -21,7 +21,8 @@ impl BlockWriterMetrics {
     ) -> Self {
         let transaction = Aggregate::new(
-            format!("{}_transaction", subgraph_id.to_string()),
+            "subgraph_transaction",
```
Could we just do `deployment`?

Jannis replied:
I'm ok with that

lutter commented:
This is all good to go from my POV

lutter commented:
Rebased to latest master and addressed an issue introduced in 27fb73c0 which made it possible to register global counters/gauges with different const labels, but you'd only ever use the one defined first since internally the global counters/gauges were kept by name, and not by the full `Desc.id`

fordN commented on graph/src/components/ethereum/adapter.rs:
```diff
@@ -580,23 +578,21 @@ impl BlockStreamMetrics {
+                "Track the number of blocks a subgraph deployment is behind the HEAD block",
+                deployment_id.as_str(),
             )
             .expect("failed to create `subgraph_blocks_behind` gauge");
```
Update error message to match the new metric name: `failed to create 'deployment_blocks_behind' gauge`.

lutter replied:
Done

fordN commented on graph/src/components/ethereum/adapter.rs:
```diff
@@ -580,23 +578,21 @@ impl BlockStreamMetrics {
+                "Track the last reverted block for a subgraph deployment",
+                deployment_id.as_str(),
             )
             .expect("Failed to create `subgraph_reverted_blocks` gauge");
```
`deployment_reverted_blocks`

lutter replied:
Done

fordN commented on core/src/subgraph/instance_manager.rs:
```diff
@@ -84,10 +84,8 @@ impl SubgraphInstanceManagerMetrics {
-                String::from(
-                    "Counts the number of subgraphs currently being indexed by the graph-node.",
-                ),
+                "subgraph_count",
```
"deployment_count"?

lutter replied:
Yep. Fixed.

fordN reviewed (approved):
👍

Jannis commented on chain/ethereum/src/network_indexer/metrics.rs:
```diff
@@ -49,246 +48,152 @@ pub struct NetworkIndexerMetrics {
         Self {
             stopwatch,
 
-            chain_head: registry
-                .new_gauge(
-                    format!("{}_chain_head", subgraph_id),
-                    "The current chain head block".into(),
-                    HashMap::new(),
-                )
-                .expect(format!("failed to register metric `{}_chain_head`", subgraph_id).as_str()),
+            chain_head: make_gauge("deployment_chain_head", "The current chain head block"),
 
-            local_head: registry
-                .new_gauge(
-                    format!("{}_local_head", subgraph_id),
-                    "The current local head block".into(),
-                    HashMap::new(),
-                )
-                .expect(format!("failed to register metric `{}_local_head`", subgraph_id).as_str()),
+            local_head: make_gauge("deployment_local_head", "The current local head block"),
```
On these two (and others related to the subgraph network): it might make sense to add a `network` label in order to be future proof for the days where one subgraph indexes across multiple networks. But for now, this is fine.

Jannis reviewed (approved):
All looking good to me!
